File: src/main/java/net/adventuregame/water/WaterRenderer.java
package net.adventuregame.water;

import com.chidozie.core.renderEngine.Loader;
import com.chidozie.core.renderEngine.WindowManager;
import net.adventuregame.entities.Camera;
import net.adventuregame.entities.Light;
import net.adventuregame.models.RawModel;
import net.adventuregame.toolbox.Maths;
import org.joml.Matrix4f;
import org.joml.Vector3f;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL13;
import org.lwjgl.opengl.GL20;
import org.lwjgl.opengl.GL30;

import java.util.List;

public class WaterRenderer {

    private static final String DUDV_MAP = "waterDUDV";
    private static final String NORMAL_MAP = "normalMap";
    private static final float WAVE_SPEED = 0.03f;

    private RawModel quad;
    private WaterShader shader;
    private WaterFrameBuffers fbos;

    private float moveFactor = 0;

    private int dudvTexture;
    private int normalTexture;


    public WaterRenderer(Loader loader, WaterShader shader, Matrix4f projectionMatrix, WaterFrameBuffers fbos) {
        this.shader = shader;
        this.fbos = fbos;
        dudvTexture = loader.loadGameTexture("DuDvmaps/" + DUDV_MAP);
        normalTexture = loader.loadGameTexture("normalmaps/" + NORMAL_MAP);
        shader.start();
        shader.connectTextureUnits();
        shader.loadProjectionMatrix(projectionMatrix);
        shader.stop();
        setUpVAO(loader);
    }


    public void render(List<WaterTile> water, Camera camera, Light sun) {
        prepareRender(camera, sun);
        for (WaterTile tile : water) {
            Matrix4f modelMatrix = Maths.createTransformationMatrix(
                    new Vector3f(tile.getX(), tile.getHeight(), tile.getZ()), 0, 0, 0,
                    WaterTile.TILE_SIZE);
            shader.loadModelMatrix(modelMatrix);
            GL11.glDrawArrays(GL11.GL_TRIANGLES, 0, quad.getVertexCount());
        }
        unbind();
    }

    private void prepareRender(Camera camera, Light sun){
        shader.start();
        shader.loadViewMatrix(camera);
        moveFactor += WAVE_SPEED * WindowManager.getFrameTimeSeconds();
        moveFactor %= 1;
        shader.loadMoveFactor(moveFactor);
        shader.loadLight(sun);
        GL30.glBindVertexArray(quad.getVaoId());
        GL20.glEnableVertexAttribArray(0);
        GL13.glActiveTexture(GL13.GL_TEXTURE0);
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.getReflectionTexture());
        GL13.glActiveTexture(GL13.GL_TEXTURE1);
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.getRefractionTexture());
        GL13.glActiveTexture(GL13.GL_TEXTURE2);
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, dudvTexture);
        GL13.glActiveTexture(GL13.GL_TEXTURE3);
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, normalTexture);
        GL13.glActiveTexture(GL13.GL_TEXTURE4);
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.getRefractionDepthTexture());
        GL11.glEnable(GL11.GL_BLEND);
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
    }

    private void unbind(){
        GL11.glDisable(GL11.GL_BLEND);
        GL20.glDisableVertexAttribArray(0);
        GL30.glBindVertexArray(0);
        shader.stop();
    }


    private void setUpVAO(Loader loader) {
        // Just x and z vectex positions here, y is set to 0 in v.shader
        float[] vertices = { -1, -1, -1, 1, 1, -1, 1, -1, -1, 1, 1, 1 };
        quad = loader.loadToVAO(vertices, 2);
    }




}

File: src/main/kotlin/net/adventuregame/water/WaterRenderer.kt
package net.adventuregame.water

import com.chidozie.core.renderEngine.Loader
import com.chidozie.core.renderEngine.WindowManager
import net.adventuregame.entity.Camera
import net.adventuregame.entity.Light
import net.adventuregame.models.RawModel
import net.adventuregame.toolbox.Maths
import org.joml.Matrix4f
import org.joml.Vector3f
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL13
import org.lwjgl.opengl.GL20
import org.lwjgl.opengl.GL30

class WaterRenderer(
    loader: Loader,
    private val shader: WaterShader,
    projectionMatrix: Matrix4f?,
    private val fbos: WaterFrameBuffers
) {
    private var quad: RawModel? = null

    private var moveFactor = 0f

    private val dudvTexture: Int
    private val normalTexture: Int


    init {
        dudvTexture = loader.loadGameTexture("DuDvmaps/" + DUDV_MAP)
        normalTexture = loader.loadGameTexture("normalmaps/" + NORMAL_MAP)
        shader.start()
        shader.connectTextureUnits()
        shader.loadProjectionMatrix(projectionMatrix!!)
        shader.stop()
        setUpVAO(loader)
    }


    fun render(water: MutableList<WaterTile>, camera: Camera, sun: Light) {
        prepareRender(camera, sun)
        for (tile in water) {
            val modelMatrix = Maths.createTransformationMatrix(
                Vector3f(tile.x, tile.height, tile.z), 0f, 0f, 0f,
                WaterTile.Companion.TILE_SIZE
            )
            shader.loadModelMatrix(modelMatrix)
            GL11.glDrawArrays(GL11.GL_TRIANGLES, 0, quad!!.vertexCount)
        }
        unbind()
    }

    private fun prepareRender(camera: Camera, sun: Light) {
        shader.start()
        shader.loadViewMatrix(camera)
        moveFactor += WAVE_SPEED * WindowManager.frameTimeSeconds
        moveFactor %= 1f
        shader.loadMoveFactor(moveFactor)
        shader.loadLight(sun)
        GL30.glBindVertexArray(quad!!.vaoId)
        GL20.glEnableVertexAttribArray(0)
        GL13.glActiveTexture(GL13.GL_TEXTURE0)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.reflectionTexture)
        GL13.glActiveTexture(GL13.GL_TEXTURE1)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.refractionTexture)
        GL13.glActiveTexture(GL13.GL_TEXTURE2)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, dudvTexture)
        GL13.glActiveTexture(GL13.GL_TEXTURE3)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, normalTexture)
        GL13.glActiveTexture(GL13.GL_TEXTURE4)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, fbos.refractionDepthTexture)
        GL11.glEnable(GL11.GL_BLEND)
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)
    }

    private fun unbind() {
        GL11.glDisable(GL11.GL_BLEND)
        GL20.glDisableVertexAttribArray(0)
        GL30.glBindVertexArray(0)
        shader.stop()
    }


    private fun setUpVAO(loader: Loader) {
        // Just x and z vectex positions here, y is set to 0 in v.shader
        val vertices = floatArrayOf(-1f, -1f, -1f, 1f, 1f, -1f, 1f, -1f, -1f, 1f, 1f, 1f)
        quad = loader.loadToVAO(vertices, 2)
    }


    companion object {
        private const val DUDV_MAP = "waterDUDV"
        private const val NORMAL_MAP = "normalMap"
        private const val WAVE_SPEED = 0.03f
    }
}
